### ISSUE TITLE
Reproducing WMT'22 Results

### ISSUE THREAD

**Issue by stephanpeitz** [marian-nmt/docmt23#1]
Hi Matt, hi Marcin,

thanks for this great contribution! 

In order to be able to replicate your WMT'22 experiments, I got a couple of questions:

- Can you share a few more hyper-parameters such as vocabulary size, learning rate, drop out etc.?
- Did you also use the <sep> token for the WMT'22 Doc(P,B_d) setup? If yes, did you sentence-align the document-level version of NewsCrawl?

Thanks a lot for your help!

Cheers,
Stephan

**Comment by stephanpeitz** [comment]
Hi, quick update:

On WMT'22 English->German, my trained Sent(P) achieves 36.5 BLEU and 59.9 COMET with following settings:

- learning rate: 0.0004
- drop out: 0.1
- warm up: 4000
- vocab. size: 50k
- num. updates: 80k
- label smoothing: 0.1
- weight decay: 0.0

These results are closer to the results in Table 5 than in Table 15. 
Do I miss something?

**Comment by mjpost** [comment]
Hi @stephanpeitz, our parameters were similar (though 32k shared vocab). Looking again at Table 15 it does seem something is amiss. I'll see if I can figure it out. Can you clarify you used the same data?

**Comment by stephanpeitz** [comment]
Hey @mjpost, thanks for your reply!

> Can you clarify you used the same data?

Sent(P) uses 295.8M lines, similar to your 297.6M lines. 
I just used `mt-data` to pull the data from WMT'22 (`mtdata get-recipe -ri wmt22-deen -o wmt22-deen`).

Sent(P,B) is still training, I'll keep you posted.

Cheers,
Stephan

**Comment by mjpost** [comment]
Well, this is embarrassing: our trainer takes the data in TSV format, and the backtrans data (with docids) was fine, but the parallel data was reversed. I had wondered about the low scores but ran out of time to investigate. Thanks for raising this to my attention. I'm retraining and will post an update soon (along with answers to your other questions).

**Comment by stephanpeitz** [comment]
Hey Matt, 

thanks for the update, much appreciated! 
No worries, these things happen! :) 

Cheers,
Stephan

**Comment by mjpost** [comment]
Okay, here are the fixed scores.

```
model		BLEU	COMET	C/Pro	GenPro
Sent(P,B)	35.8	60.6	50.8	35.3
Sent*(P,B)	-	-	73.0	34.5
Doc(P,B_d)	35.8	59.4	83.4	64.3
```

(Note that the `Sent` system sees only the payload sentence, for both scoring and generation. `Sent*`, in contrast, sees the same pseudo-document as `Doc`, but with the separator tokens removed (this only improves its performance, since it has not seen those in training). `Sent*` GenPro score is computed with `--proportion 1.2`.

and training parameters:

- learning rate: 0.0005
- drop out: 0.1
- warm up: 4000
- vocab. size: 32k
- mini-batch-fit: true
- mini-batch-words: 500000
- num. updates: 71k
- label smoothing: 0.1

**Comment by stephanpeitz** [comment]
Thanks for the corrected scores @mjpost! 
Great to see that C/Pro and G/Pro still nicely improve.

Regarding B_d: 
Did you also use the `<sep>` token for the WMT'22 Doc(P,B_d) setup?
If yes: Did you sentence-align the document-level version of NewsCrawl or did you directly use the sentence-aligned version?

Thanks!

**Comment by mjpost** [comment]
We backtranslate the German NewCrawl mono data using a normal sentence-based system and the provided sentence segmentations. When we train the Doc(P,B_d) system, we construct pseudo-documents from B_d with the `<sep>` token up to the maximum length of 250.

(I think I'm missing something from your question—NewsCrawl is mono data; what would we sentence-align to? Maybe you were wondering if we backtranslate NewCrawl _as documents_ instead of as sentences, and then aligned afterwards to restore the sentence boundaries? If that was your question, no, we don't do that.)

**Comment by stephanpeitz** [comment]
Sorry for the confusion, I meant sentence-segmented.... 

Thanks for your answer, definitely helps!

Let me close the issue for now, I will keep you posted if I'm able to reproduce all the results.

Cheers,
Stephan

**Comment by stephanpeitz** [comment]
Hi @mjpost,

couple of questions/remarks:

- Which `proportional` value have you used for scoring Sent*(P,B)?
- Which beam size did you use?
- `--json-file ContraPro/contrapro.json` should be corrected to `--json-file ext/ContraPro/contrapro.json` 

Btw, I get a G/Pro score of `34.6` with Sent(P).

Cheers,
Stephan

**Comment by mjpost** [comment]
Thanks, fixed the link.

We use a beam size of 4.

I also fixed the table of scores above after finding another error. The previous score was not `Sent` but actually `Sent*(P,B)`, as you inferred, but I had translated it *with* the separator tokens, which is unfair, since it was not trained with them. I added `Sent(P,B)` to the above (which sees only the payload sentence for both scoring and generation), and retranslated with `Sent*(P,B)` after removing those tokens. I then scored with `--proportional 1.2`. If you still believe any numbers that I produce, they are now much more comparable to what you reported :).

**Comment by stephanpeitz** [comment]
Thanks for the update! 

Let me rerun my experiments with a beam of 4.

However, it seems I get slightly higher scores with Sent*(P,B) (`54.8` GenPro).
But I'm using beam=5 and max. tokens of 1024 (instead of 250).

**Comment by stephanpeitz** [comment]
Just rerun with beam 4, still get quite high GenPro for Sent*(P,B).
Can you please double-check?
Your C/Pro score went up for Sent*(P,B) (+22.2) while for GenPro it dropped by .8.
Any idea why this is the case?

Thanks!

**Comment by mjpost** [comment]
Sure, I'll check in a bit. I wouldn't expect the beam difference to make much of a difference, nor the max token length, for just a sentence system.

What command did you use to produce your source file? Maybe we have different context sizes.

**Comment by stephanpeitz** [comment]
> What command did you use to produce your source file? Maybe we have different context sizes.

`./ext/ContraPro/bin/json2text.py -m 250  --spm spm.model --json-file ext/ContraPro/contrapro.json --correct-only > genpro.max250.en-de.tsv`
and
`cut -f 5  genpro.max250.en-de.tsv | bin/extract_sent.py -i -1  > genpro.max250.sent.en-de.en`
and
`cut -f 5  genpro.max250.en-de.tsv > genpro.max250.context.en-de.en`

**Comment by mjpost** [comment]
I looked again and don't see any obvious errors with what I've done. I just added the files here that I used in case you want to try them directly (`data/contrapo`).

C/Pro rising while G/Pro stays the same is not surprising—this reflects our finding in the paper that contrastive metrics aren't useful for discriminating document systems (corroborated on our private-data systems in [Table 6](https://arxiv.org/pdf/2304.12959.pdf)).

**Comment by stephanpeitz** [comment]
Thanks for sharing, this helped!

I just realised that I didn't add `-ms 10`. You might want to update your README as well to:
`./ext/ContraPro/bin/json2text.py -m 250  -ms 10 --spm spm.model --json-file ext/ContraPro/contrapro.json --correct-only > genpro.max250+10.en-de.tsv`

However, I still get a G/Pro score of `53.6` with following command:
`paste genpro.max250+10.en-de.tsv out-sent.genpro.max250+10.en-de.hyp | ./ext/ContraPro/bin/evaluate_tsv.py -d 1 10 -p all --proportional 1.2`

I observed that the score is quite sensitive to `proportional`.

**Comment by mjpost** [comment]
Yes, the score does change based on what you use for `proportional`. I chose 1.2 by spot-checking the output to make sure it wasn't clipping sentences. But even going up to 1.5, the numbers are quite lower than the `Doc` systems, so it seemed good enough to make what was mostly a passing point. But the better way to do this eval would be to do the alignment between the system output and either the source or the reference.

I see above you're getting 53.6 from your _sentence_ output? That seems like an error. The sentence model can only resolve items with distance 0, and should tend to pick _es_ for others. There are only 5,528 such instances, so the highest a sentence system should expect to score is about 46%.

```
cat data/contrapro/genpro.max250+10.en-de.tsv \
  | awk -F"\t" '{ if ($2 == 0 || $4 == "es") print }' \
  | wc -l
```

(That score is about what we found in the paper for our internal systems.) At the same time, my own `Sent*` score here seems low—the sentence system should be able to resolve pronouns for sentence-internal anaphora.

**Comment by stephanpeitz** [comment]
Hi, I think I found the difference: 

For `Sent(P,B)`, I did not filtered documents with more than 10 sentences. 
That means, `Sent(P,B)` is trained on lot more data than `Doc(P,B_d)`.
But it also means, `Sent*(P,B)` can do better if trained on more data?


Btw, for `Doc(P,B_d)` I achieve `67.8`, so pretty much inline with your results.

Thanks again for all your help!

Cheers,
Stephan
